### ISSUE TITLE
"Assertion `t->tick_count() < mark.ptr->key.ticks' failed to hold." in dlopen.

### ISSUE THREAD

**Issue by jeremyroman** [mozilla/rr#1609]
While trying to use rr on Chromium's "content_shell" browser (with sandbox, GPU support, and multi-process disabled to simplify things), rr has an assertion failure when I try to reverse-step.

When replaying, I was able to set a breakpoint, step forward, but then stepping back yields an assertion failure:

    Breakpoint 1, blink::FrameView::layout (this=0x274a74340010) at ../../third_party/WebKit/Source/core/frame/FrameView.cpp:889
    889         ASSERT(m_frame);
    (rr) step
    WTF::PassOwnPtr<blink::WebDataConsumerHandle>::operator! (this=0xe8d983f10475da00) at ../../third_party/WebKit/Source/wtf/RefPtr.h:72
    72          bool operator!() const { return !m_ptr; }
    (rr) reverse-step
    [FATAL ../src/ReplayTimeline.cc:432:replay_step_to_mark() errno: 0 'Success']
     (task 5147 (rec:4432) at time 14926)
     ->  Assertion `t->tick_count() < mark.ptr->key.ticks' failed to hold.
    Launch gdb with
      gdb /src/chromium/src/out/Default/content_shell
    and attach to the rr debug server with:
      target remote :5147

Attaching to the rr debug server shows the process in _dl_debug_state (the symbol the linker invokes when a shared object has just loaded, to make it easy for the debugger to break):

```
(gdb) bt 23
#0  __GI__dl_debug_state () at dl-debug.c:74
#1  0x00007f3e31e62af5 in dl_open_worker (a=a@entry=0x7f3e0dc72948) at dl-open.c:304
#2  0x00007f3e31e5dff4 in _dl_catch_error (objname=objname@entry=0x7f3e0dc72938, errstring=errstring@entry=0x7f3e0dc72940, mallocedp=mallocedp@entry=0x7f3e0dc72930, operate=operate@entry=0x7f3e31e62990 <dl_open_worker>, args=args@entry=0x7f3e0dc72948) at dl-error.c:187
#3  0x00007f3e31e623ab in _dl_open (file=0x265b44300140 "/usr/lib/x86_64-linux-gnu/nss/libnssckbi.so", mode=-2147483646, caller_dlopen=<optimized out>, nsid=-2, argc=4, argv=0x7ffedd7ad848, env=0x7ffedd7ad870) at dl-open.c:661
#4  0x00007f3e22ddd02b in dlopen_doit (a=a@entry=0x7f3e0dc72b60) at dlopen.c:66
#5  0x00007f3e31e5dff4 in _dl_catch_error (objname=0x265b443371b0, errstring=0x265b443371b8, mallocedp=0x265b443371a8, operate=0x7f3e22ddcfd0 <dlopen_doit>, args=0x7f3e0dc72b60) at dl-error.c:187
#6  0x00007f3e22ddd62d in _dlerror_run (operate=operate@entry=0x7f3e22ddcfd0 <dlopen_doit>, args=args@entry=0x7f3e0dc72b60) at dlerror.c:163
#7  0x00007f3e22ddd0c1 in __dlopen (file=file@entry=0x265b44300140 "/usr/lib/x86_64-linux-gnu/nss/libnssckbi.so", mode=<optimized out>) at dlopen.c:87
#8  0x00007f3e214adf9d in pr_LoadLibraryByPathname (name=name@entry=0x265b44300140 "/usr/lib/x86_64-linux-gnu/nss/libnssckbi.so", flags=flags@entry=26) at prlink.c:803
#9  0x00007f3e214ae4e6 in PR_LoadLibraryWithFlags (libSpec=..., flags=flags@entry=26) at prlink.c:418
#10 0x00007f3e21d22851 in loader_LoadLibInReferenceDir (referencePath=<optimized out>, referencePath@entry=0x265b443000e0 "/usr/lib/x86_64-linux-gnu/libnss3.so", name=name@entry=0x265b440d3c18 "libnssckbi.so") at secload.c:95
#11 0x00007f3e21d2292d in PORT_LoadLibraryFromOrigin (existingShLibName=existingShLibName@entry=0x7f3e22031322 "libnss3.so", staticShLibFunc=staticShLibFunc@entry=0x7f3e21f6d750 <softoken_LoadDSO>, newShLibName=newShLibName@entry=0x265b440d3c18 "libnssckbi.so") at secload.c:165
#12 0x00007f3e21f75571 in secmod_LoadPKCS11Module (mod=mod@entry=0x265b440d3b40, oldModule=oldModule@entry=0x7f3e0dc72d88) at pk11load.c:434
#13 0x00007f3e21f80a07 in SECMOD_LoadModule (modulespec=0x265b4483a7b8 "name=\"Root Certs\" library=\"libnssckbi.so\" ", parent=0x0, recurse=0) at pk11pars.c:1010
#14 0x00007f3e21f80bb9 in SECMOD_LoadUserModule (modulespec=<optimized out>, parent=<optimized out>, recurse=<optimized out>) at pk11pars.c:1098
#15 0x00007f3e2baa9b56 in crypto::(anonymous namespace)::NSSInitSingleton::LoadModule (name=0x7f3e2babddc5 <.L.str.38> "Root Certs", library_path=0x7f3e2babddd0 <.L.str.39> "libnssckbi.so", params=0x0) at ../../crypto/nss_util.cc:829
#16 0x00007f3e2baa9a56 in crypto::(anonymous namespace)::NSSInitSingleton::InitDefaultRootCerts (this=0x7f3e2bacd0e0 <crypto::(anonymous namespace)::g_nss_singleton+8>) at ../../crypto/nss_util.cc:807
#17 0x00007f3e2baa94c2 in crypto::(anonymous namespace)::NSSInitSingleton::NSSInitSingleton (this=0x7f3e2bacd0e0 <crypto::(anonymous namespace)::g_nss_singleton+8>) at ../../crypto/nss_util.cc:766
#18 0x00007f3e2baa8fe5 in base::DefaultLazyInstanceTraits<crypto::(anonymous namespace)::NSSInitSingleton>::New (instance=0x7f3e2bacd0e0 <crypto::(anonymous namespace)::g_nss_singleton+8>) at ../../base/lazy_instance.h:70
#19 0x00007f3e2baa8ec5 in base::internal::LeakyLazyInstanceTraits<crypto::(anonymous namespace)::NSSInitSingleton>::New (instance=0x7f3e2bacd0e0 <crypto::(anonymous namespace)::g_nss_singleton+8>) at ../../base/lazy_instance.h:99
#20 0x00007f3e2baa8e7d in base::LazyInstance<crypto::(anonymous namespace)::NSSInitSingleton, base::internal::LeakyLazyInstanceTraits<crypto::(anonymous namespace)::NSSInitSingleton> >::Pointer (this=0x7f3e2bacd0d8 <crypto::(anonymous namespace)::g_nss_singleton>) at ../../base/lazy_instance.h:164
#21 0x00007f3e2baa8355 in base::LazyInstance<crypto::(anonymous namespace)::NSSInitSingleton, base::internal::LeakyLazyInstanceTraits<crypto::(anonymous namespace)::NSSInitSingleton> >::Get (this=0x7f3e2bacd0d8 <crypto::(anonymous namespace)::g_nss_singleton>) at ../../base/lazy_instance.h:138
#22 0x00007f3e2baa832d in crypto::EnsureNSSInit () at ../../crypto/nss_util.cc:905
(More stack frames follow...)
```

In this case, the shared library that was loaded was libnssckbi, but I have previously seen the same symptom while loading libXcursor. My libdl is provided by eglibc-2.19-0ubuntu6.6, on Ubuntu 14.04 LTS "trusty".

It's not obvious to me what the likely cause of this is; perhaps someone has seen similar symptoms. (I haven't yet gotten around to trying to minimize a test case.)

**Comment by jeremyroman** [comment]
And the backtrace in the `rr` process:

```
(gdb) bt 20
#0  0x00007f079c0b5fa0 in __poll_nocancel () at ../sysdeps/unix/syscall-template.S:81
#1  0x00000000005bea55 in poll_socket (sock_fd=..., events=1, timeoutMs=-1) at ../src/GdbConnection.cc:239
#2  0x00000000005beaff in poll_incoming (sock_fd=..., timeoutMs=-1) at ../src/GdbConnection.cc:247
#3  0x00000000005beb4c in GdbConnection::read_data_once (this=0x7b30e50) at ../src/GdbConnection.cc:261
#4  0x00000000005bf67c in GdbConnection::read_packet (this=0x7b30e50) at ../src/GdbConnection.cc:449
#5  0x00000000005c400e in GdbConnection::get_request (this=0x7b30e50) at ../src/GdbConnection.cc:1207
#6  0x00000000005d42b9 in GdbServer::process_debugger_requests (this=0x7ffd2aa81b90, state=GdbServer::REPORT_NORMAL) at ../src/GdbServer.cc:860
#7  0x00000000005d6b72 in GdbServer::emergency_debug (t=0x2f80b80) at ../src/GdbServer.cc:1313
#8  0x00000000005f1a78 in emergency_debug (t=0x2f80b80) at ../src/log.cc:27
#9  0x00000000005f1b48 in EmergencyDebugOstream::~EmergencyDebugOstream (this=0x7ffd2aa82080, __in_chrg=<optimized out>) at ../src/log.cc:34
#10 0x0000000000658482 in ReplayTimeline::replay_step_to_mark (this=0x7ffd2aa82e18, mark=..., strategy=...) at ../src/ReplayTimeline.cc:432
#11 0x000000000065ba8d in ReplayTimeline::reverse_continue(std::function<bool (Task*)> const&, std::function<bool ()> const&) (this=0x7ffd2aa82e18, 
    stop_filter=..., interrupt_check=...) at ../src/ReplayTimeline.cc:887
#12 0x00000000005d561a in GdbServer::debug_one_step (this=0x7ffd2aa82de0, last_resume_request=...) at ../src/GdbServer.cc:1066
#13 0x00000000005d6878 in GdbServer::serve_replay (this=0x7ffd2aa82de0, flags=...) at ../src/GdbServer.cc:1278
#14 0x0000000000641823 in replay (trace_dir="", flags=...) at ../src/ReplayCommand.cc:384
#15 0x0000000000642082 in ReplayCommand::run (this=0x951db0 <ReplayCommand::singleton>, args=std::vector of length 0, capacity 1)
    at ../src/ReplayCommand.cc:478
#16 0x00000000005f3635 in main (argc=2, argv=0x7ffd2aa831b8) at ../src/main.cc:259
```

Some possibly useful data:
```
(gdb) frame 10
#10 0x0000000000658482 in ReplayTimeline::replay_step_to_mark (this=0x7ffd2aa82e18, mark=..., strategy=...) at ../src/ReplayTimeline.cc:432
432         ASSERT(t, t->tick_count() < mark.ptr->key.ticks);
(gdb) p t->tick_count()
$5 = 13121328
(gdb) p mark.ptr->key.ticks
$6 = 13121327
(gdb) p constraints
$7 = {command = RUN_CONTINUE, stop_at_time = 0, ticks_target = 13121326, stop_before_states = std::vector of length 0, capacity 0}
(gdb) p result
$8 = {status = REPLAY_CONTINUE, break_status = {task = 0x2f80b80, watchpoints_hit = std::vector of length 0, capacity 0, signal = 0, breakpoint_hit = true, 
    singlestep_complete = false, approaching_ticks_target = false, task_exit = false}, did_fast_forward = false}
```

**Comment by jeremyroman** [comment]
tl;dr: If I understand correctly, `gdb` is hard-wired to break on `_dl_debug_state` (and a handful of other symbols) so that it can trigger symbol loading when a solib is loaded, but I suspect `rr` is not expecting to stop there.

**Comment by rocallahan** [comment]
That is probably not the specific problem since rr copes just fine with those internal gdb breakpoints normally.

This error is usually a symptom of some deeper problem where rr failed to reproduce the correct state during replay so we "overshoot". These are often hard to figure out :-(.

You could retry with master where I've just fixed a replay bug that could conceivably cause this.

I assume `rr replay -a` completes successfully?

**Comment by jeremyroman** [comment]
Hmm, shame that the stack trace into `dlopen` seems a red herring, then.

`rr replay -a` also errors, in a way that is similar to #1603:

```
$ rr replay -a

rr: Warning: Your CPU frequency governor is 'ondemand'.  rr strongly
    recommends that you use the 'performance' governor.  Not using the
    'performance' governor can cause rr to be at least 2x slower
    on laptops.

    On Fedora-based systems, you can enable the 'performance' governor
    by running the following commands:

    $ sudo dnf install kernel-tools
    $ sudo cpupower frequency-set -g performance

[4419:4419:1222/161122:1559100514896:ERROR:browser_main_loop.cc(216)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
[4421:4421:1222/161122:1559100533378:FATAL:launch_posix.cc(280)] Check failed: ret == 0. : Bad file descriptor
[4419:4441:1222/161123:1559101698305:WARNING:histograms.cc(40)] Started multiple compositor clients (Browser, Renderer) in one process. Some metrics will be disabled.
[ERROR ../src/ReplaySession.cc:518:guard_overshoot() errno: 0 'Success'] 
 -> Replay diverged; target registers mismatched: 
[FATAL ../src/ReplaySession.cc:522:guard_overshoot() errno: 0 'Success'] 
 (task 22849 (rec:4441) at time 25855)
 ->  Assertion `false' failed to hold. overshot target ticks=339313819 by 9552975
Launch gdb with
  gdb /src/chromium/src/out/Default/content_shell
and attach to the rr debug server with:
  target remote :22849
```

`rr replay -a` behaves the same at the new `origin/master` (`73fe429e60f14b9cfd33f2fe8b54fcfa3c661946`); the crash I see in the target process if I attach to the emergency gdb server then is a V8 assertion failure about elapsed time while compiling a script being negative.

**Comment by rocallahan** [comment]
oooh, that sounds interesting. You might have luck tracing back the V8 assertion failure to some system call returning a different value during replay than during recording, which would point directly to an rr bug. You can use the -g option to try to attach a debugger at some event shortly before the assertion failure and step/run forward.

**Comment by jeremyroman** [comment]
Since rr has been reported to work on Chromium now, I'll assume this issue is fixed. \o/
